pubtools/_pulp/tasks/delete.py

```diff
@@ -0,0 +1,737 @@
+        return Criteria.and_(*crit)
+
+    def _get_nsvca_dict(self, module_name):
+        mod_parts = ["name", "stream", "version", "context", "arch"]
+        nsvca = module_name.split(":", 4)
+        mod_dict = {}
+        for i, p in enumerate(nsvca):
+            if mod_parts[i] == "version":
+                mod_dict[mod_parts[i]] = int(p)
+            else:
+                mod_dict[mod_parts[i]] = p
+
+        return mod_dict
+
+    def _file_search_criteria(self, file_names):
+        part_crit = []
+
+        for file_name in file_names:
+            part_crit.append(self._file_criteria(file_name))
+
+        return part_crit
+
+    def _file_criteria(self, file_name):
+        return Criteria.with_field("path", file_name)
+
+    def _rpm_search_criteria(self, rpm_names, signing_keys):
+        part_crit = []
+        signing_keys = [s.lower() for s in signing_keys] if signing_keys else [None]
+
+        for rpm_name in rpm_names:
+            part_crit.append(self._rpm_criteria(rpm_name, signing_keys))
+
+        return part_crit
+
+    def _rpm_criteria(self, filename, signing_keys):
+        return Criteria.and_(
+            Criteria.with_field("filename", filename),
+            Criteria.with_field("signing_key", Matcher.in_(signing_keys)),
+        )
+
+    def search_content(self, criteria):
+        return self.pulp_client.search_content(criteria=criteria)
+
+    def log_missing_units(self, searched_units, unit_attr, unit_type, unit_names):
+        found = []
+
+        for unit in searched_units:
+            found.append(getattr(unit, unit_attr))
+
+        missing = set(unit_names) - set(found)
+        if missing:
+            missing = sorted(list(missing))
+            LOG.warning(
+                "Requested unit(s) don't exist as %s: %s",
+                unit_type,
+                ", ".join(sorted(missing)),
+            )
+
+        LOG.info("%s unit(s) found for deletion", len(found))
+
+        return searched_units
+
+    def map_to_repo(self, units, repos, unit_attr):
+        repo_map = {}
+        unit_map = {}
+        repos = sorted(repos)
+
+        for unit in sorted(units):
+            unit_name = getattr(unit, unit_attr)
+            unit_map.setdefault(unit_name, RemoveUnitItem(unit=unit, repos=[]))
+            for repo in repos:
+                if repo not in unit.repository_memberships:
+                    LOG.warning(
+                        "%s is not present in %s",
+                        unit_name,
+                        repo,
+                    )
+                else:
+                    repo_map.setdefault(repo, []).append(unit)
+                    unit_map.get(unit_name).repos.append(repo)
+
+        missing = set(repos) - set(repo_map.keys())
+        if missing:
+            missing = ", ".join(sorted(list(missing)))
+            LOG.warning("No units to remove from %s", missing)
+
+        return repo_map, unit_map
+
+    @step("Remove RPMs")
```

Comment:
Minor - from the user's point of view, this is going to show both a "remove" and a "delete" step for each content type.  I guess users are not going to understand what's the difference without looking into the code, and it could be confusing.

Maybe this one doesn't need to be a step, or maybe it could use the more Pulp-specific technical term "Unassociate RPMs", so we'd still have the timing info but non-Pulp-expert users would just ignore it rather than being confused.